src/guest_memory.rs

```diff
@@ -166,16 +166,22 @@ pub trait GuestMemoryRegion: Bytes<MemoryRegionAddress, E = Error> {
 /// - handle cases where an access request spanning two or more GuestMemoryRegion objects.
 ///
 /// Note: all regions in a GuestMemory object must not intersect with each other.
-pub trait GuestMemory {
-    /// Type of objects hosted by the address space.
-    type R: GuestMemoryRegion;
+pub trait GuestMemory<'a> {
```

Comment:
It's so sad that `impl Trait` doesn't work with trait methods yet, because then we wouldn't have needed the lifetime parameter for `GuestMemory` (or so it seems). Do you ppl think it's significantly awkward/cumbersome to have `GuestMemory<'a>` instead of simply `<GuestMemory>`? I think there might be a way to shift the parameter over to the implementation (but it still won't be beautiful).

Reply:
It's caused me a whole day to work out this solution:(
And any better solutions are welcomed so I could get better understanding of rust:)

Reply:
I think we should hold off this pull request until we get some more experience migrating Firecracker to vm-memory (I have some initial code but nothing that I tested seriously or attempted to split into sensible commits). I'd rather avoid the risk of having an API that's not very usable; in particular if anything that has a `&GuestMemory` in Firecracker needs to become generic over the lifetime, it'd be a huge pain to do the conversion...

Reply:
Good suggestion, will give a try to migrate firecracker to the vm-memory crate:)

Reply:
I wrote a quick and somewhat dirty way of removing the generic lifetime parameter from `GuestMemory`, which is now associated with the concrete type implementing it. However, as written in the commit message, I think a lot of the awkwardness associated with passing lifetimes around can be avoided in this case by using `impl GuestMemory` instead of referring to the actual type. Anyway, if you wanna take a look, the changes are in the `nolife` branch from my fork of `vm-memory`. The commit is added on top of the latest version of Gerry's code.

Reply:
Oh snap, I forgot the `mmap` feature is conditionally compiled so I didn't test things properly :( ... pls disregard my previous message until I figure out if what I wanted to do can actually work :(

Reply:
Ahah it did seem a bit unlikely that it would compile, but I didn't trust my Rust knowledge too much...

Reply:
Ok so I found a bit more time to look into this. As long as we use associated types for the `Iterator`/`Item`, that forces us to fully declare the type as part of the implementing struct, which will require a lifetime that cannot be filled in via elision, so it has to be a parameter of either the trait or its implementation. Moreover, if we associate it with the implementation, there seems to be no way of reconciling the `iter` method from over there with the one in the trait, because the trait uses an anonymous lifetime (via `&self`) in its declaration, whereas the implementation has to link to the explicitly provided one :(

Something like the existence of an implicit `'self` lifetime (associated with the current object) would have helped, but it's not supported. Returning an `impl Trait` from `iter` would also help, but it's also not supported (yet) for traits.

That being said, there are still a couple of options that come to mind in order to avoid the lifetime parameter on `GuestMemory`:
- Return a `Box<dyn Iterator<Item=&Self::R>>` from `GuestMemory::iter`. However, this involves both heap allocation and dynamic dispatch, which might very well be a concern.
- If it's reasonable to assume that indexing over the regions in a `GuestMemory` is an efficient operation, we can add a `region(index: usize)` method to `GuestMemory` and then define an object that acts as an iterator over any kind of type implementing `GuestMemory` along the lines of
```rust
struct GuestMemoryIterator<'a, T> {
    inner: &'a T,
    next_position: usize,
}

impl<'a, T: GuestMemory> Iterator for GuestMemoryIterator<'a, T> {
    type Item = &'a T::R;
    
    fn next(&mut self) -> Option<Self::Item> {
        let res = self.inner.region(next_position)
        self.next_position += 1;
        res
    }   
}
```
If indexing is not necessarily supposed to be efficient, then we could consider some other helper function that provides a reference to the next region based on more information than just the current position. With `GuestMemoryIterator` in place, we can write `GuestMemory::iter` as
```rust 
fn iter(&self) -> GuestMemoryIterator<Self> {
    GuestMemoryIterator {
        inner: &self,
        next_position: 0,
    }
}
```
I can try and actually implement this if the assumptions make sense. Anyway, that's as much as I want to think about lifetimes for today :(

Reply:
I have tried this way but given up:(
If the GuestMemory object is implemented by array-like data structures, it's perfect because we could use a simple index to implement an iterator. But it gets harder if the GuestMemory object is implemented by tree-like data structures.

Reply:
Well, at least we tried. I'll switch my efforts towards seeing how much of a nuisance that lifetime parameter really is (maybe elision takes out most of the awkwardness). Also, I'll start looking into the implications of switching Firecracker to the new `GuestMemory` paradigm, with traits and maybe even lifetimes and all that.

Reply:
Seems the right solution to remove the lifetime parameter should be "associated type constructors". 
FYI:
https://github.com/rust-lang/rfcs/blob/master/text/1598-generic_associated_types.md
http://smallcultfollowing.com/babysteps/blog/2016/11/02/associated-type-constructors-part-1-basic-concepts-and-introduction/
http://smallcultfollowing.com/babysteps/blog/2016/11/04/associated-type-constructors-part-3-what-higher-kinded-types-might-look-like/

But "associated type constructors" is still a nightly feature, we can't rely on it yet:(
So how about adopting the solution proposed by @alexandruag as a temporary solution to remove lifetime parameter and switching to associated type constructors when it's stable?
The switch from the temporary solution to associated type constructors should almost be transparent to GuestMemory consumers and only minor changes to GuestMemory providers.

Reply:
Hi Gerry, which temporary solution are you referring to? The one from the commit link I posted doesn't work (and will take some time until the shame wears off), in case it wasn't obvious :D When I get a bit of time, I'll have a look on how awkward using that lifetime parameter for `GuestMemory` really is.